mbedtls/src/pkcs12/mod.rs

```diff
@@ -537,75 +567,172 @@ impl BERDecodable for Pkcs12PbeParams {
     }
 }
 
-// PKCS12 formats PBKDF input as BMP (UCS-16) with trailing NULL
-// See RFC 7292 Appendix B.1
-fn format_passphrase_for_pkcs12(passphrase: &str) -> Vec<u8> {
-    let mut v = Vec::with_capacity((passphrase.len() + 1) * 2);
+// See RFC 8018 Appendix A.4
+#[derive(Debug, Clone)]
+struct Pkcs12Pbes2Params {
+    kdf: Pbkdf2Params,
+    encryption_scheme: EncryptionScheme,
+}
 
-    for c in passphrase.encode_utf16().chain(core::iter::once(0)) {
-        v.extend_from_slice(&c.to_be_bytes())
+impl BERDecodable for Pkcs12Pbes2Params {
+    fn decode_ber(reader: BERReader) -> ASN1Result<Self> {
+        reader.read_sequence(|reader| {
+            let kdf = reader.next().read_sequence(|reader| {
+                let algo = reader.next().read_oid()?;
+                // Scrypt not supported
+                if algo.as_ref() != PKCS12_PBES2_PBKDF2 {
+                    Err(ASN1Error::new(ASN1ErrorKind::Invalid))?
+                }
+                read_struct_from_bytes(&reader.next().read_der()?)
+            })?;
+            let encryption_scheme = read_struct_from_bytes(&reader.next().read_der()?)?;
+            Ok(Pkcs12Pbes2Params {
+                kdf,
+                encryption_scheme,
+            })
+        })
     }
+}
 
-    v
+// See RFC 8018 Appendix A.2
+#[derive(Debug, Clone)]
+struct Pbkdf2Params {
```

Comment:
Regarding parameter validation and regardless of what PBKDF2 does internally; [RFC8018 sec. 4 "Salt and Iteration Count"](https://www.rfc-editor.org/rfc/rfc8018#section-4) provides recommendations for the length of the salt and iteration values (the latter one coming from FIPS 800-132).

Shall we put in place a mechanism to fail early on bad parameters?

Reply:
This crate is not the right place to implement policy.